core/src/main/java/org/apache/iceberg/ManifestReader.java

```diff
@@ -101,20 +100,34 @@ private String fileClass() {
 
   protected ManifestReader(
       InputFile file,
+      int specId,
       Map<Integer, PartitionSpec> specsById,
       InheritableMetadata inheritableMetadata,
       FileType content) {
     this.file = file;
     this.inheritableMetadata = inheritableMetadata;
     this.content = content;
 
+    if (specsById != null) {
+      this.spec = specsById.get(specId);
+      Preconditions.checkArgument(
```

Comment:
Why does this fail? I think it should warn and fall back to reading the spec. We don't want to needlessly introduce places where bad metadata can cause a table to be unreadable.

Reply:
Updated